File: SumToLeaf.py
#Time Complexity : O(n) 
# Space Complexity : O(h) where h is tree height for recursive stack 
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this :

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        self.finalSum = 0
        def helper(root,currSum):
            if root is None:
                return
            if (root.left is None and root.right is None):
                self.finalSum = self.finalSum + currSum * 10 + root.val
                return 
            helper(root.left,currSum * 10 + root.val)
            helper(root.right,currSum * 10 + root.val)

        helper(root,0)
        return self.finalSum

File: SumtoLeaf.java
// Time Complexity : O(n) , n for traversing over the nodes 
// Space Complexity : O(h) where h = height of the tree (max recursion calls at one point)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english : Pass value from left + right subtree to parent, and at leaf node add it to sum 

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        return helper(root,0);
    }

    private int helper(TreeNode root,int currSum){
        //base
        if(root == null) return 0;
        //logic
        int value = currSum * 10 + root.val;
        //at leaf node.
        if(root.left == null && root.right == null){
           return value;
        }

        
        return helper(root.left,value) + helper(root.right,value);
    }
}